data/transactions/logic/evalStateful_test.go

```diff
@@ -2531,3 +2518,54 @@ func TestAppAddress(t *testing.T) {
 	source = fmt.Sprintf("int 0; app_params_get AppAddress; assert; addr %s; ==;", a)
 	testApp(t, source, ep)
 }
+
+func TestBudget(t *testing.T) {
+	ep := defaultEvalParams(nil)
+	source := `
+global OpcodeBudget
+int 699
+==
+assert
+global OpcodeBudget
+int 695
+==
+`
+	testApp(t, source, ep)
+}
+
+func TestSelfMutate(t *testing.T) {
+	ep, _, ledger := makeSampleEnv()
+
+	/* In order to test the added protection of mutableAccountReference, we're
+	   going to set up a ledger in which an app account is opted into
+	   itself. That was impossible before v6, and indeed we did not have the
+	   extra mutable reference check then. */
+	ledger.NewLocals(basics.AppIndex(888).Address(), 888)
+	ledger.NewLocal(basics.AppIndex(888).Address(), 888, "hey",
```

Comment:
If the address were in the ForeignApps array, it could be modified.  This is the problem were EvalDeltas can only be written if we can use an index to name the account being modified.  So unless the account is supplied in the foreign array, it can't be allowed to be written.